cerebrium/container-images/defining-container-images.mdx

```diff
@@ -0,0 +1,288 @@
+
+```toml
+[cerebrium.dependencies.apt]
+ffmpeg = "latest"
+libopenblas-base = "latest"
+libomp-dev = "latest"
+```
+
+For teams with standardized system dependencies, text files can be used instead by adding the following to the `[cerebrium.dependencies.paths]` section:
+
+```toml
+[cerebrium.dependencies.paths]
+apt = "deps_folder/pkglist.txt"
+```
+
+Since APT packages modify the system environment, any changes to these dependencies trigger a full rebuild of the container image. This ensures system-level changes are properly integrated but means builds will take longer than when modifying Python packages alone.
+
+### Conda Packages
+
+Conda excels at managing complex system-level Python dependencies, particularly for GPU support and scientific computing:
+
+```toml
+[cerebrium.dependencies.conda]
+cuda = ">=11.7"
+cudatoolkit = "11.7"
+opencv = "latest"
+```
+
+Teams using conda environments can specify their environment file:
+
+```toml
+[cerebrium.dependencies.paths]
+conda = "conda_pkglist.txt"
+```
+
+Like APT packages, Conda packages often modify system-level components. Changes to Conda dependencies will trigger a full rebuild to ensure all binary dependencies and system libraries are correctly configured. Consider batching Conda dependency updates together to minimize rebuild time.
+
+## Commands
+
+Cerebrium provides two distinct types of commands for handling setup tasks. Understanding the difference helps optimize build times and ensures reliable app initialization.
+
+### Shell Commands
+
+Shell commands execute every time a new instance starts, making them ideal for runtime initialization and app startup:
+
+```toml
+shell_commands =  [
+    "echo 'Hello, World!'",
+    "uv venv",
+    "export SOME_ENV_VAR=value"
+]
+```
+
+Shell commands are perfect for tasks that need fresh execution on startup, such as downloading models, warming caches, or starting an app server. Since these commands run on every instance start, they should be optimized for speed to minimize startup time.
+
+### Pre-build Commands
+
+Pre-build commands execute once during image creation, making them suitable for setup tasks that can be baked into the container (Which should also improve cold start performance):
+
+```toml
+pre_build_commands = [
+    "curl -o /usr/local/bin/pget -L 'https://github.com/replicate/pget/releases/download/v0.6.2/pget_linux_x86_64' && chmod +x /usr/local/bin/pget"
+]
+```
+
+Pre-build commands are ideal for compiling code, building assets, or installing build-time dependencies. Since these commands affect the container image, changes to pre-build commands trigger a full rebuild. Group related commands together to minimize rebuild frequency.
+
+## Custom Docker Base Images
+
+The choice of base image fundamentally shapes how an app runs in Cerebrium. While the default Debian slim image works well for most Python apps, some scenarios require specialized base images.
+
+### Selecting Base Images
```

Comment:
remove this title, just move the paragraph up under the custom dicker base images title